### PR TITLE
Adding `cartesian` as a Table join_type

### LINKED ISSUE #8220: Allow table cartesian join with no common column
From user Igor Bukanov on Astropy feature request page:

Currently table join operations requires to have a common column in the tables. That prevents to form a Cartesian product of tables. The later is handy if one wants to calculate something based on data from unrelated tables.

### PR DESCRIPTION
**Pull request by bsipocz**
This is part of my yearly pilgrimage to table operations land, very simply following what was suggested in the issue.

To close #8220

### CODE REVIEW

**Comment by taldcroft** [comment]
Nice, and clever idea.  In a perfect world this would be done in the Cython `_np_utils.join_inner()`, but I looked at that code and I don't really remember how it works, it's all a little intricate.  I think your solution is going to be just as fast.

The main comment I have is that the in-place addition of a column can leave the input table(s) modified if there is an exception during the join.  One option would be (only in the case of a cartesian join), to make light copies of `left` and `right`.  With #8404 this should be reasonably fast and memory efficient.  Something like `left = left.copy(copy_data=False)`.  This makes a key-only copy of the corresponding meta, so also light and fast.

File: astropy/table/tests/test_operations.py [hunk 1]
```diff
@@ -560,15 +560,23 @@ def test_mixin_functionality(self, mixin_cols):
                 assert ('join requires masking' in str(err.value) or
                         'join unavailable' in str(err.value))
 
+    def test_cartesian_join(self, operation_table_type):
+        self._setup(operation_table_type)
+        t1 = self.t1
+        t2 = self.t2
+
+        t12 = table.join(t1, t2, join_type='cartesian')
+        assert len(t12) == len(t1) * len(t2)
```
Comment (taldcroft) [inline astropy/table/tests/test_operations.py]: You should confirm that the table elements make sense.  Like at this point:
```
t1['idx'] = 0
t2['idx'] = 0
t12_outer = table.join(t1, t2, join_type='outer', keys='idx')
del t12_outer['idx']
# check that t12 column names and values match t12_outer
```

File: astropy/table/tests/test_operations.py [hunk 2]
```diff
@@ -560,15 +560,23 @@ def test_mixin_functionality(self, mixin_cols):
...
+
```
Comment (taldcroft) [inline astropy/table/tests/test_operations.py]: Add a test that original table is not affected by an exception:
```
t1 = Table([[1, 2]], meta={'a': [1]})
t2 = Table([[3, 4]], meta={'a': 'b'})
with pytest.raises(MergeConflictError):
    table.join(t1, t2, join_type='cartesian', metadata_conflicts='error')
assert t1.colnames == ['col0']
assert t2.colnames == ['col0']
```

**Comment by bsipocz** [comment]
Oh, this was on my list, but nevertheless thanks for finishing up, there are plenty other on that list 😉

**Comment by taldcroft** [comment]
@bsipocz - you are the next victim on my finishing Table PRs sprint.  I think this is done now, have a look.  Fingers crossed on CI.

**Comment by bsipocz** [comment]
@taldcroft - all your changes look good to me, but I cannot approve as I opened the PR. So it needs to be either you or @mhvk who merges this.

**Pull request merged**